hq_superset/models.py

```diff
@@ -11,3 +17,71 @@ class DataSetChange:
     def __post_init__(self):
         if 'doc_id' not in self.data:
             raise TypeError("'data' missing required key: 'doc_id'")
+
+
+class HQClient(db.Model, OAuth2ClientMixin):
+    __tablename__ = 'hq_oauth_client'
+
+    domain = db.Column(db.String(255), primary_key=True)
+
+    def check_client_secret(self, client_secret):
+        return check_password_hash(self.client_secret, client_secret)
+
+    def revoke_tokens(self):
+        tokens = db.session.execute(
+            db.select(Token).filter_by(client_id=self.client_id, revoked=False)
+        ).all()
+        for token, in tokens:
+            token.revoked = True
+            db.session.add(token)
+        db.session.commit()
+
+    @classmethod
+    def get_by_domain(cls, domain):
+        return db.session.query(HQClient).filter_by(domain=domain).first()
+
+    @classmethod
+    def get_by_client_id(cls, client_id):
+        return db.session.query(HQClient).filter_by(client_id=client_id).first()
+
+    @classmethod
+    def create_domain_client(cls, domain: str):
+        client_secret = str(uuid.uuid4())
+        client = HQClient(
+            domain=domain,
+            client_id=str(uuid.uuid4()),
+            client_secret=generate_password_hash(client_secret),
+        )
+        client.set_client_metadata({"grant_types": ["client_credentials"]})
+
+        db.session.add(client)
+        db.session.commit()
+
+        return client.client_id, client_secret
+
+
+class Token(db.Model):
+    __tablename__ = 'hq_oauth_token'
+
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    client_id = db.Column(db.String(40), nullable=False, index=True)
+    token_type = db.Column(db.String(40))
+    access_token = db.Column(db.String(255), nullable=False, unique=True)
+    revoked = db.Column(db.Boolean, default=False)
+    issued_at = db.Column(db.DateTime, default=datetime.utcnow)
+    expires_at = db.Column(db.DateTime)
+    scope = db.Column(db.String(255))
+
+    @property
+    def domain(self):
+        client = HQClient.get_by_client_id(self.client_id)
+        return client.domain
+
+    def is_expired(self):
+        return datetime.utcnow() > self.expires_at
+
+    def is_revoked(self):
+        return self.revoked
+
+    def get_scope(self):
+        return self.scope
```

Comment:
It's not obvious to me why you created a method that returns a property that is already public, and it doesn't look like a subclass will override this method. Do we need it?

Reply:
The `get_scope` is a method that the `require_oauth` decorator is looking for on the Token class, so it has to be defined as such. I'll add a comment and make that clear.